### PR TITLE
[net] add support for DNS Stateful Operations (DSO)

### PR DESCRIPTION
**Pull request by abtink**
This commit adds support for DNS Stateful Operations (DSO) as
specified in RFC 8490.

It adds `platform/dso_transport.hpp` header file which defines the
platform APIs/callbacks for DSO transport layer (e.g., DSN-over-TLS
or DNS-over-TCP).

The `Dso` module handles establishing connection with a peer, acting
either as a DSO client or server, establishing a DSO session over a
connection, and then sending and processing DSO request, response,
and unidirectional messages (including support for DSO TLV formats).
The `Dso` module also manages the session life cycle and timeouts,
namely the "Inactivity" and "Keep Alive" timeouts (including sending
and processing of Keep Alive messages when needed). It also handles
adding encryption padding before sending a message. It implements the
padding policy "Random-Block-Length Padding" from RFC 8467.

This commit also adds a detailed unit test `test_dso` covering the
behavior (including corner cases) of the `Dso` implementation. The
unit test provides an implementation of the DSO platform APIs which
emulate the DSO transport layer. It also includes a simplified alarm
platform implementation (emulating timers and allowing time to
advance in the unit test). These allow the unit test to cover more
complicated situations and behaviors (timeouts, failures, etc).

------

_Quick background on this:_ 
DSO is used by SRP replication (SRPL) and/or DNS push.

### CODE REVIEW

File: include/openthread/platform/dso_transport.h [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ * For DNS-over-TLS, the transport layer MUST listen on port 853 and follow RFC 7858.
+ *
+ * While listening is enabled, if a connection request is received, the `otPlatDsoAccept()` callback MUST be called.
+ *
+ * @param[in] aInstance    The OpenThread instance
+ * @param[in] aEnable      TRUE to start listening, FALSE to stop listening.
+ *
+ */
+void otPlatDsoEnableListening(otInstance *aInstance, bool aEnabled);
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/platform/dso_transport.h:79]: DSO can operate on different transports, over TCP or over TLS.

SRPL seems to require DNS-over-TLS which is on port 853.  

The DSO implementation is currently aimed for SRPL as its first use-case but I guess it can be extended to handle other use-cases when/if needed.

> Another potential issue I can see is that platform will need to use the same listening socket for both DSO and non-DSO DNS over TCP connections.

That's correct, but I guess it is possible to do so. This API is just requesting to start listening for connection requests from others on DSO transport (on a single socket or may be even on multiple sockets). Platform may have already a socket open for other operations (say for DNS-over-TCP) and then reuse the existing open socket for DSO as well.

File: include/openthread/platform/dso_transport.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+ * peer. The platform reports the outcome by invoking `otPlatDsoHandleConnected()` callback on success or
+ * `otPlatDsoHandleDisconnected()` callback on failure.
+ *
+ * @param[in] aInstance      The OpenThread instance.
+ * @param[in] aPeerSockAddr  The socket address (IPv6 address and port number) of the peer requesting connection.
+ *
+ * @returns A pointer to the `otPlatDsoConnection` to use if to accept, or `NULL` if to reject.
+ *
+ */
+extern otPlatDsoConnection *otPlatDsoAccept(otInstance *aInstance, const otSockAddr *aPeerSockAddr);
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/platform/dso_transport.h:98]: > When will otPlatDsoAccept be called? Will it be called when the TCP connection is accepted? Or will it be called when the DSO session is established?

`otPlatDsoAccept()` is a callback from DSO transport platform layer. 
- DSO transport layer uses this callback to notify the DSO core module about a a connection request from a peer and get informed whether or not the DSO module wants to accept and use this connection _for DSO operation_.
- It is not tied to DSO session establishment. DSO core module will take care of DSO session establishment. 
- It is tied to the notion of connection request at DSO transport layer (e.g., TCP connection request for DNS-over-TCP/TLS).
- If `otPlatDsoAccept()` rejects a connection request, it just means that DSO use-case does not want this connection for DSO use. Transport/platform layer can still continue to establish connection and use the connection/socket for other purposes (say DNS-over TCP). 

> IIUC, A DNS over TCP connection can be non-DSO connection for a period of time, transmitting normal DNS over TCP messages, until the client choose to establish the DSO session.

That's possible and supported in the current code. We can establish a connection with a peer and not start with DSO session establishment on it for a while. It is up to the user of DSO module (next layer above DSO, e.g. SRPL) when/how to start the DSO session establishment.

**Comment by abtink** [comment]
New push 
- Rebases the PR on `main`
- Adds new capability in `SendRequestMessage()` for caller to specify a response timeout per message (if not given default value will be used) - unit test is also updated to validate this functionality
- Changes `Accept` callback to include `Instance`
- Includes some callback renames (`SignalConnected` -> `HandleConnected`) and typo fixes.

File: include/openthread/platform/dso_transport.h (hunk 1 above)
Comment (jwhui) [inline include/openthread/platform/dso_transport.h:79]:
```suggestion
void otPlatDsoEnableListening(otInstance *aInstance, bool aEnable);
```

File: include/openthread/platform/dso_transport.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    OT_PLAT_DSO_DISCONNECT_MODE_GRACEFULLY_CLOSE, ///< Gracefully close the connection.
+    OT_PLAT_DSO_DISCONNECT_MODE_FORCIBLY_ABORT,   ///< Forcibly abort the connection.
+} otPlatDsoDisconnectMode;
+
+/**
+ * This function requests a connection to be disconnected.
+ *
+ * After calling this function, the DSO platform implementation MUST no longer maintain `aConnection` pointer (platform
+ * MUST not call any callbacks using this `Connection` pointer anymore). In particular, calling `otPlatDsoDisconnect()`
+ * MUST not trigger the callback `otPlatDsoHandleDisconnected()`.
```
Comment (jwhui) [inline include/openthread/platform/dso_transport.h:177]:
```suggestion
 * After calling this function, the DSO platform implementation MUST NOT maintain `aConnection` pointer (platform
 * MUST NOT call any callbacks using this `Connection` pointer anymore). In particular, calling `otPlatDsoDisconnect()`
 * MUST NOT trigger the callback `otPlatDsoHandleDisconnected()`.
```

File: include/openthread/platform/dso_transport.h [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+ *
+ */
+void otPlatDsoDisconnect(otPlatDsoConnection *aConnection, otPlatDsoDisconnectMode aMode);
+
+/**
+ * This function is a callback from the platform layer to indicate that peer closed/aborted the connection or the
+ * connection establishment failed (e.g., peer rejected a connection request).
+ *
+ * After calling this function, the DSO platform implementation MUST no longer maintain `aConnection` pointer (platform
+ * MUST not call any callbacks using this `Connection` pointer anymore).
```
Comment (jwhui) [inline include/openthread/platform/dso_transport.h:190]:
```suggestion
 * After calling this function, the DSO platform implementation MUST NOT maintain `aConnection` pointer (platform
 * MUST NOT call any callbacks using this `Connection` pointer anymore).
```

File: src/core/common/time.hpp [hunk 5]
```diff
@@ -238,15 +238,15 @@ class Time : public Unequatable<Time>
      * @returns The number of milliseconds.
      *
      */
-    static uint32_t SecToMsec(uint32_t aSeconds) { return aSeconds * 1000u; }
+    static uint32_t constexpr SecToMsec(uint32_t aSeconds) { return aSeconds * 1000u; }
```
Comment (jwhui) [inline src/core/common/time.hpp:241]: Document param.

File: src/core/common/time.hpp [hunk 6]
```diff
@@ -238,15 +238,15 @@ class Time : public Unequatable<Time>
...
 
     /**
      * This static method converts a given number of milliseconds to seconds.
      *
      * @returns The number of seconds.
      *
      */
-    static uint32_t MsecToSec(uint32_t aMilliseconds) { return aMilliseconds / 1000u; }
+    static uint32_t constexpr MsecToSec(uint32_t aMilliseconds) { return aMilliseconds / 1000u; }
```
Comment (jwhui) [inline src/core/common/time.hpp:249]: Document param.

File: src/core/net/dns_dso.cpp [hunk 7]
```diff
@@ -0,0 +1,1537 @@
...
+    // returned in `aPrimaryTlvType` (set to `Tlv::kReservedType`
+    // (value zero) when `kErrorNotFound`).
+
+    Error error = kErrorNotFound;
+    Tlv   tlv;
+
+    aPrimaryTlvType = Tlv::kReservedType;
+
+    SuccessOrExit(aMessage.Read(aMessage.GetOffset(), tlv));
+    VerifyOrExit(aMessage.GetOffset() + tlv.GetSize() <= aMessage.GetLength(), error = kErrorParse);
```
Comment (jwhui) [inline src/core/net/dns_dso.cpp:779]: Do we need to worry about potential overflow of `offset + size`?

File: src/core/net/dns_dso.cpp [hunk 8]
```diff
@@ -0,0 +1,1537 @@
...
+        }
+        break;
+
+    case kStateConnectedButSessionless:
+    case kStateEstablishingSession:
+    case kStateSessionEstablished:
+        if (mPendingRequests.HasAnyTimedOut(aNow))
+        {
+            // If server sends no response to a request, client
+            // waits for 30 section (`kResponeTimeout) after which
```
Comment (jwhui) [inline src/core/net/dns_dso.cpp:1268]:
```suggestion
            // waits for 30 section (`kResponseTimeout`) after which
```

File: src/core/net/dns_dso.cpp [hunk 9]
```diff
@@ -0,0 +1,1537 @@
...
+const char *Dso::Connection::DisconnectReasonToString(DisconnectReason aReason)
+{
+    static const char *const kDisconnectReasonStrings[] = {
+        "FailedToConnect",         // (0) kReasonFailedToConnect
+        "ResponseTimeout",         // (1) kReasonResponseTimeout
+        "PeerDoesNotSupportDso",   // (2) kReasonPeerDoesNotSupportDso
+        "PeerClosed",              // (3) kReasonPeerClosed
+        "PeerAborted",             // (4) kReasonPeerAborted
+        "InactivityTimeout",       // (5) kReasonInactivityTimeout
+        "KeepAliveTimout",         // (6) kReasonKeepAliveTimout
```
Comment (jwhui) [inline src/core/net/dns_dso.cpp:1362]:
```suggestion
        "KeepAliveTimeout",        // (6) kReasonKeepAliveTimeout
```

File: src/core/net/dns_dso.cpp [hunk 10]
```diff
@@ -0,0 +1,1537 @@
...
+            // without any messages being sent or received, the server
+            // considers the client delinquent and aborts the connection.
+
+            if (IsClient())
+            {
+                IgnoreError(SendKeepAliveMessage());
+            }
+            else
+            {
+                Disconnect(kForciblyAbort, kReasonKeepAliveTimout);
```
Comment (jwhui) [inline src/core/net/dns_dso.cpp:1307]:
```suggestion
                Disconnect(kForciblyAbort, kReasonKeepAliveTimeout);
```

File: src/core/net/dns_dso.hpp [hunk 11]
```diff
@@ -0,0 +1,979 @@
...
+         */
+        enum DisconnectReason : uint8_t
+        {
+            kReasonFailedToConnect,         ///< Failed to connect (e.g., peer did not accept or timed out).
+            kReasonResponseTimeout,         ///< Response timeout (no response from peer after `kResponseTimeout`).
+            kReasonPeerDoesNotSupportDso,   ///< Peer does not support DSO.
+            kReasonPeerClosed,              ///< Peer closed the connection gracefully.
+            kReasonPeerAborted,             ///< Peer forcibly aborted the connection.
+            kReasonInactivityTimeout,       ///< Connection closed or aborted due to Inactivity timeout.
+            kReasonKeepAliveTimout,         ///< Connection closed due to Keep Alive timeout.
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:238]:
```suggestion
            kReasonKeepAliveTimeout,        ///< Connection closed due to Keep Alive timeout.
```

File: src/core/net/dns_dso.hpp [hunk 12]
```diff
@@ -0,0 +1,979 @@
...
+         * This class defines the callback functions used by a `Connection`.
+         *
+         */
+        class Callbacks
+        {
+            friend class Connection;
+
+        public:
+            /**
+             * This callback signals that the connection is established (entering state `kStateConnectedButSessionless).
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:254]:
```suggestion
             * This callback signals that the connection is established (entering state `kStateConnectedButSessionless`).
```

File: src/core/net/dns_dso.hpp [hunk 13]
```diff
@@ -0,0 +1,979 @@
...
+             * The offset in @p aMessage is set to point to the start of the DSO TLVs. DSO module only reads and
+             * validates the first TLV (primary TLV) from the message. It is up to the callback implementation to parse
+             * and validate the rest of the TLVs in the message.
+             *
+             * @param[in] aConnection      A reference to the connection.
+             * @param[in] aMessageId       The message ID of the received request.
+             * @param[in] aMessage         The received message. Message offset is set to the start of the TLVs.
+             * @param[in] aPrimaryTlvType  The primary TLV type.
+             *
+             * One of the following `Error` values MUST be returned:
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:302]: nit: I don't think this note is necessary

File: src/core/net/dns_dso.hpp [hunk 14]
```diff
@@ -0,0 +1,979 @@
...
+            ProcessRequestMessage        mProcessRequestMessage;
+            ProcessUnidirectionalMessage mProcessUnidirectionalMessage;
+            ProcessResponseMessage       mProcessResponseMessage;
+        };
+
+        /**
+         * This constructor initializes a `Connection` instance.
+         *
+         * The `kDefaultTimeout` will be used for @p aInactivityTimeout and @p aKeepAliveInterval. The
+         * @p aKeepAliveInterval MUST not be less than `kMinKeepAliveInterval`.
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:411]:
```suggestion
         * @p aKeepAliveInterval MUST NOT be less than `kMinKeepAliveInterval`.
```

File: src/core/net/dns_dso.hpp [hunk 15]
```diff
@@ -0,0 +1,979 @@
...
+        Dns::Header::Response mRetryDelayErrorCode;
+        DisconnectReason      mDisconnectReason;
+    };
+
+    /**
+     * This callback function is used by DSO module to determine whether or not to accept a connection request from a
+     * peer.
+     *
+     * The function MUST return a non-null `Connection` pointer if the request is to be accepted. The returned
+     * `Connection` instance MUST be in `kStateDisconnetd`. The DSO module will take the ownership of the `Connection`
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:849]:
```suggestion
     * `Connection` instance MUST be in `kStateDisconnected`. The DSO module will take the ownership of the `Connection`
```

File: src/core/net/dns_dso.hpp [hunk 16]
```diff
@@ -0,0 +1,979 @@
...
+     * @param[in] aInstance      The OpenThread instance.
+     * @param[in] aPeerSockAddr  The peer socket address.
+     *
+     * @returns A pointer to the `Connection` to use if to accept, or `nullptr` if to reject the connection request.
+     *
+     */
+    typedef Connection *(*AcceptHandler)(Instance &aInstance, const Ip6::SockAddr &aPeerSockAddr);
+
+    /**
+     * This constructor initializes the `Dso` module
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:864]:
```suggestion
     * This constructor initializes the `Dso` module.
```

File: src/core/net/dns_dso.hpp [hunk 17]
```diff
@@ -0,0 +1,979 @@
...
+     *
+     */
+    explicit Dso(Instance &aInstance);
+
+    /**
+     * This method starts listening for DSO connection requests from peers.
+     *
+     * Once a connection request (from a peer) is received, the `Dso` module will invoke the `AppectHandler` to
```
Comment (jwhui) [inline src/core/net/dns_dso.hpp:872]:
```suggestion
     * Once a connection request (from a peer) is received, the `Dso` module will invoke the `AcceptHandler` to
```

File: src/core/net/dns_dso.cpp (hunk 7 above)
Comment (abtink) [reply src/core/net/dns_dso.cpp:779]: I think it is covered and handled properly.
-  The DSO TLVs use `uint16_t` as length (`Tlv::GetLenhth()` returns `uint16_t`).
- But `Tlv::GetSize()` actually returns `uint32_t` (to ensure size is correct even when length is `UINT16_MAX`).
- So the check here is done in `uint32_t` which should be fine.

File: src/core/net/dns_dso.hpp [hunk 18]
```diff
@@ -0,0 +1,979 @@
...
+namespace Dns {
+
+using ot::Encoding::BigEndian::HostSwap16;
+using ot::Encoding::BigEndian::HostSwap32;
+
+extern "C" otPlatDsoConnection *otPlatDsoAccept(otInstance *aInstance, const otSockAddr *aPeerSockAddr);
+
+extern "C" void otPlatDsoHandleConnected(otPlatDsoConnection *aConnection);
+extern "C" void otPlatDsoHandleReceive(otPlatDsoConnection *aConnection, otMessage *aMessage);
+extern "C" void otPlatDsoHandleDisconnected(otPlatDsoConnection *aConnection, otPlatDsoDisconnectMode aMode);
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/dns_dso.hpp:69]: Yes, it is supported.

File: tests/unit/test_dso.cpp [hunk 19]
```diff
@@ -0,0 +1,1240 @@
...
+    {
+        Log("   Request rejected");
+        ExitNow();
+    }
+
+    Log("   Request accepted");
+    VerifyOrQuit(peerConn->GetState() == Connection::kStateConnecting);
+
+    Log("   Signalling `Connected` on peer connection (%s)", peerConn->GetName());
+    otPlatDsoHandleConnected(peerConn);
```
[the remark this answers is not in the document]
Comment (abtink) [reply tests/unit/test_dso.cpp:409]: Yes. It is supported.

File: include/openthread/platform/dso_transport.h [hunk 20]
```diff
@@ -0,0 +1,202 @@
...
+ * free the message when no longer needed.
+ *
+ * The @p aMessage MUST contain the DNS message (starting with DNS header) and not include the length field that may
+ * be included in TCP/TLS exchange.
+ *
+ * @param[in] aConnection   The connection on which the message was received.
+ * @param[in] aMessage      The received message.
+ *
+ */
+extern void otPlatDsoHandleReceive(otPlatDsoConnection *aConnection, otMessage *aMessage);
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/platform/dso_transport.h:160]: It is correct that `otMessage` is opaque type, but we have a lot of public OT APIs in `message.h` to allocate/append/free, etc for platform to use (similar to `Message` which does not expose its internal definitions and provide methods). We provide:
- `otMessageAppend()` to append byte to an `otMessage`
- `otMessageRead()` and `otMessageWrite()`, `otMessageFree()`, ...
- `otIp6NewMessage` to allocate a new message.

File: src/core/net/dns_dso.cpp [hunk 21]
```diff
@@ -0,0 +1,1537 @@
...
+    case kStateConnecting:
+    case kStateEstablishingSession:
+        break;
+    };
+
+exit:
+    return;
+}
+
+Message *Dso::Connection::NewMessage(void)
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/dns_dso.cpp:150]: We have `otIp6NewMessage()`

File: include/openthread/platform/dso_transport.h [hunk 22]
```diff
@@ -0,0 +1,202 @@
...
+extern "C" {
+#endif
+
+/**
+ * This structure represents a DSO connection.
+ *
+ * It is an opaque struct (the platform implementation only deals with pointers to this struct).
+ *
+ */
+typedef struct otPlatDsoConnection otPlatDsoConnection;
```
[the remark this answers is not in the document]
Comment (abtink) [reply include/openthread/platform/dso_transport.h:55]: The model to use by a platform implementation is for it to have its own `struct/class` to maintain a socket/connection which includes the `otPlatDsoConnection *` along with any other data it wants. This is more flexible.

One way to think of this is that the `otPlatDsoConnection` is kind of like a file descriptor, a token we give to platform so it can later use to refer to the same connection in other API calls.

File: src/core/net/dns_dso.cpp [hunk 23]
```diff
@@ -0,0 +1,1536 @@
...
+}
+
+void Dso::Connection::Connect(void)
+{
+    OT_ASSERT(mState == kStateDisconnected);
+
+    Init();
+    mIsServer = false;
+    Get<Dso>().mClientConnections.Push(*this);
+    MarkAsConnecting();
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/dns_dso.cpp:163]: Yes. It is definitely intentional. We don't need to signal all state changes. There are only certain events in specific situations where we need to signal to the next layer.
   
The `SignalAnyStateChange` mechanism is used to make sure we signal the changes and invoke the `Connection` related callback *at the end and after processing and making all changes*. This is very important or else we can create code re-entry issues. There are comments in the code in couple places (`HandleReceive()` or `SendMessage()`) which try to clarify and emphasize this:
```c++
    // Signal any state changes. This is done at the very end when the
    // `SendMessage()` is fully processed (all state and local
    // variables are updated) to ensure that we do not have any
    // reentrancy issues (e.g., if the callback signalling state
    // change triggers another tx).

    SignalAnyStateChange();
```

File: src/core/net/dns_dso.cpp [hunk 24]
```diff
@@ -0,0 +1,1537 @@
...
+    return Get<MessagePool>().Allocate(Message::kTypeOther, sizeof(Dns::Header),
+                                       Message::Settings(Message::kPriorityNormal));
+}
+
+void Dso::Connection::Connect(void)
+{
+    OT_ASSERT(mState == kStateDisconnected);
+
+    Init();
+    mIsServer = false;
```
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/dns_dso.cpp:161]: Good idea. Changed in new push.

File: include/openthread/platform/dso_transport.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (superwhd) [reply include/openthread/platform/dso_transport.h:55]: That makes sense. Thanks for the explanation!

**Review by superwhd — APPROVED** [review]
LGTM :+1:

File: src/core/net/dns_dso.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply src/core/net/dns_dso.cpp:151]: I figure otIp6NewMessage() can create a buffer. 

However, a TCP DNS message can easily exceed the capacity of one message buffer?
Instead of receiving all data of a DNS message, which may require significant buffer size, we may better need to create a message buffer when the length is received, and continue to append new data to the otMessage when more data is received?

Do we have APIs for constructing a large otMessage to hold incoming data stream?

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 

with small questions

File: src/core/net/dns_dso.cpp (hunk 21 above)
[the remark this answers is not in the document]
Comment (abtink) [reply src/core/net/dns_dso.cpp:151]: `otMessage()` does not technically have a size limit. It automatically grows as new data is appended/written to it (e.g. using `otMessageAppend()`).
-  The underlying `Message` implementation will allocate and chain new buffer blocks as new data is added, but that is not visible to the users of `otMessage`. 
-  I guess the limit is technically the overall buffer pool size, but a single `otMessage` can potentially use all buffers from pool and entire pool size, so it can grow very large (also message length is `uint16_t` which limit the length to  ~65K).

**Review by jwhui — APPROVED** [review]
